File: internal/runner/deps.go
package runner

import (
	"context"
)

// waitForDeps blocks until all dependencies are ready. Returns false
// if any dependency failed or the context was cancelled.
func (r *Runner) waitForDeps(ctx context.Context, name string, deps []string) bool {
	for _, dep := range deps {
		depProc := r.cfg.Processes[dep]

		depMode := "launch"
		if len(depProc.ExitCodes) > 0 {
			depMode = "exit with expected code"
		} else if depProc.Readiness.HasReadiness() {
			depMode = "pass readiness check"
		}

		r.mux.SystemLog("%s waiting for %s to %s...", name, dep, depMode)

		bc := r.deps[dep]

		select {
		case <-bc.ch:
			if bc.state != depReady {
				r.mux.SystemLog("%s: dependency %s failed to become ready", name, dep)

				return false
			}
		case <-ctx.Done():
			r.mux.SystemLog("%s cancelled while waiting for %s", name, dep)

			return false
		}
	}

	return true
}

// signalDepResult broadcasts a dependency result. Only the first call
// per process has effect (guarded by sync.Once). Closing the channel
// wakes all waiting goroutines simultaneously.
func (r *Runner) signalDepResult(name string, state depState) {
	bc := r.deps[name]
	bc.once.Do(func() {
		bc.state = state
		close(bc.ch)
	})
}

File: src/runner/deps.rs
use crate::config::ReadyWhen;

use super::Shared;

/// How a dependency resolved. Each process broadcasts exactly one terminal
/// state to everything waiting on it.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum DepState {
    /// Not yet resolved.
    Pending,
    /// Ready: launched, exited with an expected code, or passed readiness.
    Ready,
    /// Failed to become ready.
    Failed,
}

impl Shared {
    /// Blocks until all dependencies are ready. Returns `false` if any
    /// dependency failed or a shutdown was requested.
    pub(crate) async fn wait_for_deps(&self, name: &str, deps: &[String]) -> bool {
        for dep in deps {
            if !self.wait_for_dep(name, dep).await {
                return false;
            }
        }
        true
    }

    /// Broadcasts a dependency result. Only the first call per process has
    /// any effect; later calls are ignored, waking all waiters at once.
    pub(crate) fn signal_dep_result(&self, name: &str, state: DepState) {
        if let Some(tx) = self.deps.get(name) {
            tx.send_if_modified(|current| {
                if *current == DepState::Pending {
                    *current = state;
                    true
                } else {
                    false
                }
            });
        }
    }

    /// Blocks until one dependency resolves. Returns `false` if it failed or a
    /// shutdown was requested; an unknown dependency is treated as satisfied.
    async fn wait_for_dep(&self, name: &str, dep: &str) -> bool {
        let dep_mode = self
            .service(dep)
            .map_or(ReadyWhen::Launched, |svc| svc.process().ready_when())
            .verb();
        self.system
            .debug(format!("{name} waiting for {dep} to {dep_mode}..."));

        let Some(tx) = self.deps.get(dep) else {
            return true;
        };
        let mut rx = tx.subscribe();

        tokio::select! {
            () = self.token.cancelled() => {
                self.system.debug(format!("{name} cancelled while waiting for {dep}"));
                false
            }
            result = rx.wait_for(|s| *s != DepState::Pending) => {
                let ready = result.is_ok_and(|state| *state == DepState::Ready);
                // A manual stop already explained itself via a per-tab note.
                if !ready && !self.is_stopped(dep) {
                    self.system
                        .warn(format!("{name}: dependency {dep} failed to become ready"));
                }
                ready
            }
        }
    }
}
